api/src/main/java/marquez/db/DbRetention.java

```diff
@@ -0,0 +1,112 @@
+                  table, retentionDays));
+      // Determine whether any rows met the retention policy.
+      if (totalNumOfRowsDeleted == 0) {
+        log.info("No rows older than '{}' days in '{}' table.", retentionDays, table);
+        continue;
+      }
+      log.info("Successfully deleted '{}' rows from '{}' table.", totalNumOfRowsDeleted, table);
+    }
+  }
+
+  /** Returns the number of {@code row}s deleted by the provided retention {@code query}. */
+  private static int executeDbRetentionQuery(
+      @NonNull Jdbi jdbi, @NonNull final String retentionQuery) throws DbRetentionException {
+    try {
+      return jdbi.withHandle(
+          handle -> {
+            final Stopwatch stopwatch = Stopwatch.createStarted();
+            int totalNumOfRowsDeleted = 0; // Keep count of rows deleted.
+            boolean hasMoreRows = true;
+            while (hasMoreRows) {
+              // Apply batch deletes (for improved performance) on rows in table with a creation
+              // time > date allowed in retention query.
+              final List<UUID> rowsDeleted =
+                  handle.createQuery(retentionQuery).mapTo(UUID.class).stream().toList();
+              final int numOfRowsDeleted = rowsDeleted.size();
+              totalNumOfRowsDeleted += numOfRowsDeleted;
+              hasMoreRows = !rowsDeleted.isEmpty();
+            }
+            // Measure the time elapsed to execute retention query, then return number of rows
+            // deleted.
+            stopwatch.stop();
+            log.debug(
+                "Deleted '{}' rows in '{}' secs using retention policy: {}",
+                totalNumOfRowsDeleted,
+                stopwatch.elapsed().toSeconds(),
+                retentionQuery);
+            return totalNumOfRowsDeleted;
+          });
+    } catch (StatementException errorOnDbRetentionQuery) {
+      log.error("Failed to apply retention query: {}", retentionQuery);
+      // Propagate throwable up the stack.
+      throw new DbRetentionException(errorOnDbRetentionQuery.getCause());
+    }
+  }
+
+  /** Build retention {@code query} for the provided {@code table} and {@code retentionDays}. */
+  private static String buildQuery(
+      @NonNull String queryTemplate, @NonNull String table, final int retentionDays) {
+    final String pk = pkInQueryFor(table);
+    final String timestamp = timestampInQueryFor(table);
+    return queryTemplate.formatted(
+        table, pk, pk, table, timestamp, retentionDays, BULK_DELETE_BATCH_SIZE, pk);
+  }
+
+  /** Returns {@code pk} for the provided {@code table}. */
+  private static String pkInQueryFor(@NonNull final String table) {
+    return LINEAGE_EVENTS.equals(table) ? Columns.RUN_UUID : Columns.ROW_UUID;
+  }
+
+  /** Returns {@code timestamp} for the provided {@code table}. */
+  private static String timestampInQueryFor(@NonNull final String table) {
+    return LINEAGE_EVENTS.equals(table) ? Columns.EVENT_TIME : Columns.CREATED_AT;
```

Comment:
If we delete dataset that was created more than `X` days ago, we also delete its' lineage that doesn't meet the retention period defined. Aren't we? 

In other words: with a retention of 7 days, a dataset that was created 8 days ago, and is filled daily with new data, should not be deleted. Otherwise, we lose lineage before the retention time defined. 

I am not sure if this is working like that but I think it definitely makes sense to write a test that assures that.

Reply:
It would make sense to me to start with deleting old job/dataset versions and delete job/datasets when they have no versions anymore?

Reply:
@pawel-big-lebowski, @julienledem: Very valid. The retention query should be smarter about when a job or dataset was last updated before applying a retention policy. In the case for a job: _when was the last recent run_? For a dataset, it may have been created past the allowed retention days, but _are new versions are still being created by job runs?_

I can easily make the adjustments to the retention query, but will keep the [`retention` config](https://github.com/MarquezProject/marquez/blob/feature/db-retention/marquez.example.yml#L31) the same with the plan to expand configuration options similar to what was proposed in https://github.com/MarquezProject/marquez/issues/1271.